lib/services/local/generic/generic_wrapper.go

```diff
@@ -0,0 +1,102 @@
+func NewServiceWrapper[T types.ResourceMetadata](
+	backend backend.Backend,
+	resourceKind string,
+	backendPrefix string,
+	marshalFunc MarshalFunc[T],
+	unmarshalFunc UnmarshalFunc[T]) (*ServiceWrapper[T], error) {
+
+	cfg := &ServiceConfig[resourceMetadataAdapter[T]]{
+		Backend:       backend,
+		ResourceKind:  resourceKind,
+		BackendPrefix: backendPrefix,
+		MarshalFunc: func(w resourceMetadataAdapter[T], option ...services.MarshalOption) ([]byte, error) {
+			return marshalFunc(w.resource, option...)
+		},
+		UnmarshalFunc: func(bytes []byte, option ...services.MarshalOption) (resourceMetadataAdapter[T], error) {
+			r, err := unmarshalFunc(bytes, option...)
+			return newResourceMetadataAdapter(r), trace.Wrap(err)
+		},
+	}
+	service, err := NewService[resourceMetadataAdapter[T]](cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &ServiceWrapper[T]{service: service}, nil
+}
+
+// ServiceWrapper is an adapter for Service that makes it usable with RFD 153-style resources,
+// which implement types.ResourceMetadata.
+//
+// Not all methods from Service are exported, in the effort to reduce the API complexity
+// as well as adhere to the guidance from RFD 153, but additional methods may be exported in the future as needed.
+type ServiceWrapper[T types.ResourceMetadata] struct {
+	service *Service[resourceMetadataAdapter[T]]
+}
+
+// UpsertResource upserts a resource.
+func (s ServiceWrapper[T]) UpsertResource(ctx context.Context, resource T) (T, error) {
+	adapter, err := s.service.UpsertResource(ctx, newResourceMetadataAdapter(resource))
+	return adapter.resource, trace.Wrap(err)
+}
+
+// UpdateResource updates an existing resource.
+func (s ServiceWrapper[T]) UpdateResource(ctx context.Context, resource T) (T, error) {
+	adapter, err := s.service.UpdateResource(ctx, newResourceMetadataAdapter(resource))
+	return adapter.resource, trace.Wrap(err)
+}
+
+// CreateResource creates a new resource.
+func (s ServiceWrapper[T]) CreateResource(ctx context.Context, resource T) (T, error) {
+	adapter, err := s.service.CreateResource(ctx, newResourceMetadataAdapter(resource))
+	return adapter.resource, trace.Wrap(err)
+}
+
+// GetResource returns the specified resource.
+func (s ServiceWrapper[T]) GetResource(ctx context.Context, name string) (resource T, err error) {
+	adapter, err := s.service.GetResource(ctx, name)
+	return adapter.resource, trace.Wrap(err)
+}
+
+// DeleteResource removes the specified resource.
+func (s ServiceWrapper[T]) DeleteResource(ctx context.Context, name string) error {
+	return trace.Wrap(s.service.DeleteResource(ctx, name))
+}
+
+// ListResources returns a paginated list of resources.
+func (s ServiceWrapper[T]) ListResources(ctx context.Context, pageSize int, pageToken string) ([]T, string, error) {
+	adapters, nextToken, err := s.service.ListResources(ctx, pageSize, pageToken)
+	out := make([]T, 0, len(adapters))
+	for _, adapter := range adapters {
+		out = append(out, adapter.resource)
+	}
```

Comment:
Suggestion:

```suggestion
	for i, adapter := range adapters {
		out[i] = adapter.resource
	}
```

Reply:
My bad, it should include the line `out := make([]T, len(adapters))`. I'll just resolve.